lib/components/chat_history/twitch/message.dart

```diff
@@ -277,4 +272,29 @@ class TwitchMessageWidget extends StatelessWidget {
           ));
     });
   }
+
+  List<InlineSpan> processText(String message, BttvEmoteModel bttvEmotes,
+      StyleModel styleModel, TextStyle linkStyle, TextStyle tagStyle) {
+    final List<InlineSpan> children = [];
+    final tokenArray = message.split(" ");
+
+    for (var token in tokenArray) {
+      final emote =
+          bttvEmotes.globalEmotes[token] ?? bttvEmotes.channelEmotes[token];
+      if (emote != null) {
+        final url = "https://cdn.betterttv.net/emote/${emote.id}/1x";
+        children.add(WidgetSpan(
+            alignment: PlaceholderAlignment.middle,
+            child:
+                Image(image: NetworkImage(url), height: styleModel.fontSize)));
+      } else {
+        children.addAll(parseText(
```

Comment:
I think the biggest issue is that this creates one `Widget` per token which can be a lot of widgets on screen. We can improve this by iterating and manually finding tokens. Actually, this algorithm is not too different from the one @chungwwei wrote for tags although maybe a little more complicated :)

```
var lastParsedStart = 0;
for (var start = 0; start < message.length; ) {
  final end = message.indexOf(" ", start) + 1;
  final token = end == 0 ? message.substring(start) : message.substring(start, end);
  if (bttvEmotes. ... [token.trim()] != null) {
    // this token is an emote. parse off the text from the last parsed start.
    children.addAll(parseText(message.substring(lastParsedStart, start)), ...);
    // then add the widget for the emote.
    children.add(WidgetSpan(...));
    // then move the last parsed start pointer.
    lastParsedStart = end;
  } else {
    // this token is not an emote, so try the next token.
    start = end;
  }
}
if (lastParsedStart != message.length) {
  // add the tokens that are after the last emote.
  children.addAll(parseText(message.substring(lastParsedStart)), ...);
}
```

Reply:
Would this function make sense as a top level function just like the other parse functions in this file?

Reply:
Eh I think it's fine to put in the top level if you want. It can be made private by prefixing the name with an underscore. Static functions and top-level functions are basically the same in dart, I think it's just up to preference.